### DISCUSSION TITLE
Performance problem wit streaming response

### DISCUSSION THREAD

**Discussion by Enormoss** [fastapi/fastapi#7212] [Questions] [answered]
Hi I noticed slow response using StreamingResponse.

compared to Flask it is much more slower.

Flask code runnig under flask server:
```
from flask import Flask, Response

app = Flask("app")

@app.route("/")
def stream():
    def generate():
        while True:
            yield "test string\r\n"

    return Response(generate(), headers={"Content-Type": "text/plain"})

if __name__ == "__main__":
    app.run(host="0.0.0.0", port=60580)
```
testing with curl I get:
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 10.0M    0 10.0M    0     0   419k      0 --:--:--  0:00:24 --:--:--  380
```
FastAPI code running under uvicorn:

```
from fastapi import FastAPI
from fastapi.responses import StreamingResponse


app = FastAPI()

@app.get("/")
def stream():
    def generate():
        while True:
            yield "test  string\r\n"

    response = StreamingResponse(generate(), media_type="text/plain")

    return response
```
I get
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1203k    0 1203k    0     0  35558      0 --:--:--  0:00:34 --:--:-- 37239
```
you can compare Dload speed which is very slow comapred to Flask.

**Comment by Enormoss** [comment]
Environment is the same, Python3.8, CentOS Linux release 8.1.1911 (Core)

FastAPI 0.61.1
Uvicorn 0.12.2
starlette 0.13.6

**Comment by ycd** [comment, marked as answer]
Can you try comparing a fixed amount?

Compare this with an equal Flask Code to this. 

```python
from fastapi import FastAPI
from fastapi.responses import StreamingResponse

app = FastAPI()


async def test_string_generator():
    for _ in range(100000):
        yield "test  string\r\n"


@app.get("/")
async def main():
    return StreamingResponse(test_string_generator())
```

```zsh
time curl --silent --output /dev/null http://127.0.0.1:8000

Out: 0,00s user 0,01s system 35% cpu 0,043 total
```

**Comment by Enormoss** [comment]
Amazing! That's it, now it is  a little faster than Flask and even CPU loads drops.
Thank you

**Comment by scriptator** [comment]
> Can you try comparing a fixed amount?

According to my investigations with that code, what makes the real difference is making the generator function `async`. Omit that and the example takes 7 seconds instead of 0.1 seconds on my machine. The reason is probably that StreamingResponse can only handle async iterators and if you pass in a non-async one it needs to offload that to a thread pool which seems to have quite some overhead.

**Comment by iashraful** [comment]
**Moral of the issue:** We must use async generator/iterator to get the best performance. I was stuck a day with same kind of problem. Finally, I rewrite the my small chunk of code with async generator and it worked like champ. ❤️
